rules/ast/ast.go

```diff
@@ -749,55 +756,185 @@ func labelsEqual(labels1, labels2 clientmodel.Metric) bool {
+
+		if node.opType == And {
+			value = ls.Value
+			keep = true
+		} else {
+			// arithmetic operation
+			if node.matchType == MatchOTO && added[fp] {
+				panic("many to one matching must be explicit")
+			}
+			// account for potentially swapped sidedness.
+			vl, vr := ls.Value, rs.Value
+			if node.matchType == MatchOTM {
+				vl, vr = vr, vl
 			}
+			value, keep = evalVectorBinop(node.opType, vl, vr)
 		}
-		return result
-	} else if node.lhs.Type() == VectorType && node.rhs.Type() == VectorType {
-		lhs := node.lhs.(VectorNode).Eval(timestamp)
-		rhs := node.rhs.(VectorNode).Eval(timestamp)
-		for _, lhsSample := range lhs {
-			for _, rhsSample := range rhs {
-				if labelsEqual(lhsSample.Metric.Metric, rhsSample.Metric.Metric) {
-					value, keep := evalVectorBinop(node.opType, lhsSample.Value, rhsSample.Value)
-					if keep {
-						lhsSample.Value = value
-						if node.opType.shouldDropMetric() {
-							lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-						}
-						result = append(result, lhsSample)
-					}
-				}
+
+		if keep {
+			ns := &Sample{
+				Metric:    node.resultMetric(ls, rs),
+				Value:     value,
+				Timestamp: timestamp,
+			}
+			result = append(result, ns)
+			added[fp] = true
+		}
+	}
+
+	// add all remaining samples in the rhs in an OR operation if they
+	// have not been matched up with a lhs sample.
+	if node.opType == Or {
+		for fp, rs := range rm {
+			if !added[fp] {
+				rs.Metric = node.resultMetric(rs, nil)
+				result = append(result, rs)
+			}
+		}
+	}
+	return result
+}
+
+// resultMetric returns the metric for the given sample based on the vector
+// binary operation and the matching options.
+func (node *VectorArithExpr) resultMetric(ls, rs *Sample) clientmodel.COWMetric {
+	if len(node.matchBy) == 0 {
+		if node.opType.shouldDropMetric() {
+			ls.Metric.Delete(clientmodel.MetricNameLabel)
+		}
+		return ls.Metric
+	}
+
+	m := clientmodel.Metric{}
+	for _, ln := range node.matchBy {
+		m[ln] = ls.Metric.Metric[ln]
+	}
+	for _, ln := range node.includeLabels {
+		// Try to find a label/value pair on either side, starting to the left.
+		// If the label is present on both sides an error is thrown.
+		v, ok := ls.Metric.Metric[ln]
+		if rs != nil {
+			vr, okr := rs.Metric.Metric[ln]
+			if ok && okr {
+				panic("label available on both sides only allowed in by clause")
+			}
+			if okr {
+				v, ok = vr, okr
 			}
 		}
-		return result
+		if ok {
+			m[ln] = v
+		}
+	}
+	return clientmodel.COWMetric{false, m}
+}
+
+// fingerprint calculates the fingerprint for the given metric based on the matching
+// options for the binary operation.
+func (node *VectorArithExpr) fingerprint(cm clientmodel.COWMetric) clientmodel.Fingerprint {
+	if len(node.matchBy) == 0 {
+		m := cm.Metric.Clone()
+		delete(m, clientmodel.MetricNameLabel)
+		return m.Fingerprint()
 	}
-	panic("Invalid vector arithmetic expression operands")
+
+	m := clientmodel.Metric{}
+	for _, ln := range node.matchBy {
```

Comment:
Here we don't need a full fingerprint with prior copy of map keys. Since the relevant label names are constant and known, we just need to write the label *values* into a hash, like we do in `labelsToGroupingKey`: https://github.com/prometheus/prometheus/blob/master/rules/ast/ast.go#L374-L382.